### PR TITLE
Update CI to Julia 1.8.1

### PR DESCRIPTION
**Pull request by ilopezgp**
## Issue number (if applicable)

### CODE REVIEW

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
bors try-

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
bors try-

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
There are some issues with IO to the netCDF files in this new NCDatasets version. E.g.,

```
┌ Error: Failed to write array of dimension () as val_mse_full_var to NetCDF file. Expected array of dimension ().
└ @ CalibrateEDMF.NetCDFIO /central/home/ilopezgo/CalibrateEDMF.jl/src/NetCDFIO.jl:243
```

This is probably related to #401 .

**Comment by ilopezgp** [comment]
bors try

File: src/NetCDFIO.jl [hunk 1]
```diff
@@ -224,35 +224,44 @@ function close_files(diags::NetCDFIO_Diags)
     close(diags.root_grp)
 end
 
-"""Adds a given field to an existing NetCDF Dataset."""
+"""
+    add_field(diags::NetCDFIO_Diags, var_name; dims, group, type)
+
+Add a field with some `var_name` and `type` to some `group` in an existing NetCDF Dataset. The dataset is assume closed.
```
Comment (ilopezgp) [inline src/NetCDFIO.jl:230]:
```suggestion
Add a field with some `var_name` and `type` to some `group` in an existing NetCDF Dataset. The dataset is assumed closed.
```

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
bors try

**Comment by ilopezgp** [comment]
> Note: IO is significantly simplified in [CliMA/TurbulenceConvection.jl#1214](https://github.com/CliMA/TurbulenceConvection.jl/pull/1214). We should consider doing a similar refactoring effort.

Isn't the problem with that PR that it is significantly slower? I prefer speed over simplicity.

**Comment by haakon-e** [comment]
> > Note: IO is significantly simplified in [CliMA/TurbulenceConvection.jl#1214](https://github.com/CliMA/TurbulenceConvection.jl/pull/1214). We should consider doing a similar refactoring effort.
> 
> Isn't the problem with that PR that it is significantly slower? I prefer speed over simplicity.

I think Charlie is still working on it.

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try-

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try-

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try

**Comment by haakon-e** [comment]
bors try

**Comment by ilopezgp** [comment]
@haakon-e is there a specific outcome you are looking for?

**Comment by ilopezgp** [comment]
bors try-

**Comment by haakon-e** [comment]
It appears that `bomex inv julia_par` fails in every bors try since the last update to the branch, with basically the same error message:

<details><summary>ProcessExitedException(8)</summary>
<p>

Example from [here](https://buildkite.com/clima/calibrateedmf-ci/builds/2274#018380eb-303d-46fc-9d39-785a3dd49317).

```
[ Info: Running EK updates for 10 iterations
--
  | [ Info: iter = 1
  | Worker 8 terminated.
  | Unhandled Task ERROR: EOFError: read end of file
  | Stacktrace:
  | [1] (::Base.var"#wait_locked#680")(s::Sockets.TCPSocket, buf::IOBuffer, nb::Int64)
  | @ Base ./stream.jl:941
  | [2] unsafe_read(s::Sockets.TCPSocket, p::Ptr{UInt8}, nb::UInt64)
  | @ Base ./stream.jl:950
  | [3] unsafe_read
  | @ ./io.jl:759 [inlined]
  | [4] unsafe_read(s::Sockets.TCPSocket, p::Base.RefValue{NTuple{4, Int64}}, n::Int64)
  | @ Base ./io.jl:758
  | [5] read!
  | @ ./io.jl:760 [inlined]
  | [6] deserialize_hdr_raw
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/messages.jl:167 [inlined]
  | [7] message_handler_loop(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/process_messages.jl:172
  | [8] process_tcp_streams(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/process_messages.jl:133
  | [9] (::Distributed.var"#103#104"{Sockets.TCPSocket, Sockets.TCPSocket, Bool})()
  | @ Distributed ./task.jl:484
  | Worker 10 terminated.
  | Unhandled Task ERROR: IOError: read: connection reset by peer (ECONNRESET)
  | Stacktrace:
  | [1] wait_readnb(x::Sockets.TCPSocket, nb::Int64)
  | @ Base ./stream.jl:410
  | [2] (::Base.var"#wait_locked#680")(s::Sockets.TCPSocket, buf::IOBuffer, nb::Int64)
  | @ Base ./stream.jl:944
  | [3] unsafe_read(s::Sockets.TCPSocket, p::Ptr{UInt8}, nb::UInt64)
  | @ Base ./stream.jl:950
  | [4] unsafe_read
  | @ ./io.jl:759 [inlined]
  | [5] unsafe_read(s::Sockets.TCPSocket, p::Base.RefValue{NTuple{4, Int64}}, n::Int64)
  | @ Base ./io.jl:758
  | [6] read!
  | @ ./io.jl:760 [inlined]
  | [7] deserialize_hdr_raw
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/messages.jl:167 [inlined]
  | [8] message_handler_loop(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/process_messages.jl:172
  | [9] process_tcp_streams(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/process_messages.jl:133
  | [10] (::Distributed.var"#103#104"{Sockets.TCPSocket, Sockets.TCPSocket, Bool})()
  | @ Distributed ./task.jl:484
  | From worker 9:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 6:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 2:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 4:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 11:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 5:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 7:	[ Info: Length of g_scm (full, pca): (180, 13)
  | From worker 3:	[ Info: Length of g_scm (full, pca): (180, 13)
  | ERROR: LoadError: ProcessExitedException(8)
  | Stacktrace:
  | [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
  | @ Base ./task.jl:871
  | [2] wait()
  | @ Base ./task.jl:931
  | [3] wait(c::Base.GenericCondition{ReentrantLock})
  | @ Base ./condition.jl:124
  | [4] take_buffered(c::Channel{Any})
  | @ Base ./channels.jl:416
  | [5] take!(c::Channel{Any})
  | @ Base ./channels.jl:410
  | [6] take!(::Distributed.RemoteValue)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:726
  | [7] remotecall_fetch(f::Function, w::Distributed.Worker, args::String; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:461
  | [8] remotecall_fetch(f::Function, w::Distributed.Worker, args::String)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:454
  | [9] remotecall_fetch(f::Function, id::Int64, args::String; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:492
  | [10] remotecall_fetch
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/remotecall.jl:492 [inlined]
  | [11] remotecall_pool(rc_f::typeof(remotecall_fetch), f::Function, pool::WorkerPool, args::String; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/workerpool.jl:123
  | [12] remotecall_pool
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/workerpool.jl:120 [inlined]
  | [13] #remotecall_fetch#200
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/workerpool.jl:229 [inlined]
  | [14] remotecall_fetch
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/workerpool.jl:229 [inlined]
  | [15] #208#209
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/workerpool.jl:274 [inlined]
  | [16] #208
  | @ /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/workerpool.jl:274 [inlined]
  | [17] (::Base.var"#929#934"{Distributed.var"#208#210"{Distributed.var"#208#209#211"{WorkerPool, typeof(scm_eval_train)}}})(r::Base.RefValue{Any}, args::Tuple{String})
  | @ Base ./asyncmap.jl:100
  | [18] macro expansion
  | @ ./asyncmap.jl:234 [inlined]
  | [19] (::Base.var"#945#946"{Base.var"#929#934"{Distributed.var"#208#210"{Distributed.var"#208#209#211"{WorkerPool, typeof(scm_eval_train)}}}, Channel{Any}, Nothing})()
  | @ Base ./task.jl:484
  | Stacktrace:
  | [1] (::Base.var"#939#941")(x::Task)
  | @ Base ./asyncmap.jl:177
  | [2] foreach(f::Base.var"#939#941", itr::Vector{Any})
  | @ Base ./abstractarray.jl:2774
  | [3] maptwice(wrapped_f::Function, chnl::Channel{Any}, worker_tasks::Vector{Any}, c::Vector{String})
  | @ Base ./asyncmap.jl:177
  | [4] wrap_n_exec_twice
  | @ ./asyncmap.jl:153 [inlined]
  | [5] #async_usemap#924
  | @ ./asyncmap.jl:103 [inlined]
  | [6] #asyncmap#923
  | @ ./asyncmap.jl:81 [inlined]
  | [7] pmap(f::Function, p::WorkerPool, c::Vector{String}; distributed::Bool, batch_size::Int64, on_error::Nothing, retry_delays::Vector{Any}, retry_check::Nothing)
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/pmap.jl:126
  | [8] pmap(f::Function, p::WorkerPool, c::Vector{String})
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/pmap.jl:99
  | [9] pmap(f::Function, c::Vector{String}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/pmap.jl:156
  | [10] pmap(f::Function, c::Vector{String})
  | @ Distributed /central/software/julia/1.8.1/share/julia/stdlib/v1.8/Distributed/src/pmap.jl:156
  | [11] macro expansion
  | @ /central/scratch/esm/slurm-buildkite/calibrateedmf-ci/2274/calibrateedmf-ci/integration_tests/julia_parallel_test.jl:65 [inlined]
  | [12] macro expansion
  | @ ./timing.jl:262 [inlined]
  | [13] top-level scope
  | @ /central/scratch/esm/slurm-buildkite/calibrateedmf-ci/2274/calibrateedmf-ci/integration_tests/julia_parallel_test.jl:61
  | in expression starting at /central/scratch/esm/slurm-buildkite/calibrateedmf-ci/2274/calibrateedmf-ci/integration_tests/julia_parallel_test.jl:60
  | 🚨 Error: The command exited with status 1
  | user command error: exit status 1

```

</p>
</details>

One test that is quite similar to `Bomex inv Julia_par` is `TRMM stretch + micro`. That test failed 7/8 of the most recent tries (success on the most recent try before you cancelled the bors). Apart from running `TRMM` instead of `Bomex, the possibly relevant changes to the config are:

Bomex does:
`config["Δt"] = PiecewiseConstantDecay(1.0, 5)` in `get_process_config()` and
```
config["save_tc_output"] = true  # save sims
config["save_tc_iterations"] = [1, 4]  # save subset of iterations
```
in `get_output_config()`.

I doubt the time step algorithm may be relevant, but perhaps there is an error in the method for saving output that I'm not able to parse from the relevant stacktrace. 

For both cases, whenever the test fails, it's always somewhere in [`pmap(scm_eval_train, versions)`](https://github.com/CliMA/CalibrateEDMF.jl/blob/main/integration_tests/julia_parallel_test.jl#L65)

For other tests, such as `SCT1 Julia parallel` however, for all 3/8 times it failed, it was the validation set: [`pmap(scm_eval_validation, versions)`](https://github.com/CliMA/CalibrateEDMF.jl/blob/main/integration_tests/julia_parallel_test.jl#L66). That test uses UKI instead of EKI, as the major difference from the TRMM test, as far as I can see (and also doesn't save output).

Across the last 8 bors tries I did, of tests that failed at least one time, the `total number of failures / 8 tries` are:
`create system image`: 1/8
`Bomex inv julia_par`: 8/8
`SCT1 julia parallel`: 3/8
`heterogeneous batching`: 4/8
`TRMM stretch + micro`: 7/8
`grid search`: 6/8
`TurbulenceConvectionUtils`: 4/8
`ReferenceModels`: 4/8
`ReferenceStats`: 2/8
`Pipeline`: 2/8 
`NetCDFIO`: 2/8

**Comment by ilopezgp** [comment]
Thanks for the detailed response! I would say: let's run the Bomex inv integration test manually on the cluster. If it passes, then let's merge

**Comment by ilopezgp** [comment]
I am also hoping that the `pmap` issues are resolved after the revamp of the HPC cluster. We will see.

**Comment by ilopezgp** [comment]
@haakon-e I am going to try to merge this.

**Comment by haakon-e** [comment]
> @haakon-e I am going to try to merge this.

Please do! Thanks.

**Comment by ilopezgp** [comment]
@haakon-e The errors were OOM, I just explicitly requested 6G per CPU and now the Bomex integration test is passing.

**Comment by ilopezgp** [comment]
bors ping

**Pull request merged**